source/common/common/compiled_string_map.h

```diff
@@ -0,0 +1,180 @@
+ * `x-prefix-babana`
+ * `x-prefix-apple`
+ * `x-prefix-pineapple`
+ * `x-prefix-barana`
+ * `x-prefix-banaka`
+ *
+ * A standard front-first trie looking for `x-prefix-banana` would walk
+ * 7 nodes through the tree, first for `x`, then for `-`, etc.
+ *
+ * This structure first jumps to matching length, eliminating in this
+ * example case apple and pineapple.
+ * Then the "best split" node is on
+ *   `x-prefix-banana`
+ *               ^
+ * so the first node has 3 non-miss branches, n, b and r for that position.
+ * Down that n branch, the "best split" is on
+ *   `x-prefix-banana`
+ *                 ^
+ * which has two branches, n or k.
+ * Down the n branch is the leaf node (only `x-prefix-banana` remains) - at
+ * this point a regular string-compare checks if the key is an exact match
+ * for the string node.
+ */
+template <class Value> class CompiledStringMap {
+  // While it is usual to take a string_view by value, in this
+  // performance-critical context with repeatedly passing the same
+  // value, passing it by reference benchmarks out slightly faster.
+  using FindFn = std::function<Value(const absl::string_view&)>;
+
+public:
+  using KV = std::pair<absl::string_view, Value>;
+  /**
+   * Returns the value with a matching key, or the default value
+   * (typically nullptr) if the key was not present.
+   * @param key the key to look up.
+   */
+  Value find(const absl::string_view& key) const {
+    if (key.size() >= table_.size() || table_[key.size()] == nullptr) {
+      return {};
+    }
+    return table_[key.size()](key);
+  };
+  /**
+   * Construct the lookup table. This can be a somewhat slow multi-pass
+   * operation if the input table is large.
+   * @param initial a vector of key->value pairs. This is taken by value because
+   *                we're going to modify it. If the caller still wants the original
+   *                then it can be copied in, if not it can be moved in.
+   *                Note that the keys are string_views - the base string data must
+   *                exist for the duration of compile(). The leaf nodes take copies
+   *                of the key strings, so the string_views can be invalidated once
+   *                compile has completed.
+   */
+  void compile(std::vector<KV> initial) {
+    if (initial.empty()) {
+      return;
+    }
+    std::sort(initial.begin(), initial.end(),
+              [](const KV& a, const KV& b) { return a.first.size() < b.first.size(); });
+    size_t longest = initial.back().first.size();
+    table_.resize(longest + 1);
+    auto range_start = initial.begin();
+    // Populate the sub-nodes for each length of key that exists.
+    while (range_start != initial.end()) {
+      // Find the first key whose length differs from the current key length.
+      // Everything in between is keys with the same length.
+      auto range_end =
+          std::find_if(range_start, initial.end(), [l = range_start->first.size()](const KV& e) {
```

Comment:
I'm not a fan of 'l' as a variable name :)

can we have 'last'?

Reply:
len. :)